src/main/java/bio/terra/service/resourcemanagement/OneProjectPerResourceSelector.java

```diff
@@ -0,0 +1,81 @@
+package bio.terra.service.resourcemanagement;
+
+import bio.terra.model.BillingProfileModel;
+import bio.terra.service.dataset.Dataset;
+import bio.terra.service.resourcemanagement.exception.GoogleProjectNamingException;
+import bio.terra.service.resourcemanagement.google.GoogleResourceConfiguration;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Component;
+import com.google.common.hash.Hashing;
+
+import java.util.UUID;
+
+@Component
+@Profile({"terra", "google"})
+public class OneProjectPerResourceSelector implements DataLocationSelector {
+    private final GoogleResourceConfiguration resourceConfiguration;
+    private static final String GS_PROJECT_PATTERN = "[a-z0-9\\-]{6,30}";
+    /**
+     * *Borrowed from terra-resource-buffer*
+     * The size of project when generating random characters. Choose size as 8 based on AoU's
+     * historical experience, increase if 8 is not enough for a pool's naming.
+     */
+    static final int RANDOM_ID_SIZE = 8;
+
+    @Autowired
+    public OneProjectPerResourceSelector(GoogleResourceConfiguration resourceConfiguration) {
+        this.resourceConfiguration = resourceConfiguration;
+    }
+
+    @Override
+        public String projectIdForDataset() throws GoogleProjectNamingException {
+        return getNewProjectId();
+    }
+
+    @Override
+    public String projectIdForSnapshot() throws GoogleProjectNamingException {
+        return getNewProjectId();
+    }
+
+    @Override
+    public String projectIdForFile(Dataset dataset, BillingProfileModel billingProfile)
+        throws GoogleProjectNamingException {
+        UUID sourceDatasetBillingProfileId = dataset.getProjectResource().getProfileId();
+        UUID requestedBillingProfileId = UUID.fromString(billingProfile.getId());
+        if (sourceDatasetBillingProfileId.equals(requestedBillingProfileId)) {
+            return dataset.getProjectResource().getGoogleProjectId();
+        } else {
+            return getNewProjectId();
+        }
+
+    }
+
+    @Override
+    public String bucketForFile(Dataset dataset, BillingProfileModel billingProfile)
+        throws GoogleProjectNamingException {
+        return projectIdForFile(dataset, billingProfile) + "-bucket";
+    }
+
```

Comment:
@myessail - I haven't gotten a chance to think through how this change works with your PR.